### ISSUE TITLE
Geography support

### ISSUE THREAD

**Issue by EvanCarroll** [sfackler/rust-postgres#891]
I have a table that I've created like this,

```sql
CREATE TABLE foo (
    coords geography(POINT)
);
```

When I try to select from it too a `geo_types::Point`, I get

> thread 'tokio-runtime-worker' panicked at 'error retrieving column coords: error deserializing column 3: cannot convert between the Rust type `geo_types::point::Point<f64>` and the Postgres type `geography`', /home/ecarroll/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-postgres-0.7.5/src/row.rs:151:25

It would be nice if this functionality was provided with `with-geo-types-0_7`

**Comment by EvanCarroll** [comment]
I see what this crate provides, it provides the the ability to go from the native [geometric type](https://www.postgresql.org/docs/current/datatype-geometric.html#id-1.5.7.16.5) to a geo_type. This is probably not very useful for a PostGIS user which is stores `geography`. Though you can convert with `::geometry::point`

Ie,

    SELECT geog::geometry::point
    FROM ( VALUES (ST_MakePoint(0,0)::geography) ) AS t(geog);

**Comment by jbg** [comment]
@EvanCarroll in case you haven't seen it, you may find the [postgis crate](https://docs.rs/postgis) useful.

**Comment by jbg** [comment]
> But ideally this type would just support PostGIS types regardless. If rust-postgres's type crate is interested in support PostGIS I could write a patch for it.

Does rust-postgres currently support any postgres types that are not either core postgres types or provided by extensions included with the core postgres distribution? Support for third-party extensions seems better in separate crates IMO.

**Comment by EvanCarroll** [comment]
> Does rust-postgres currently support any postgres types that are not either core postgres types or provided by extensions included with the core postgres distribution? Support for third-party extensions seems better in separate crates IMO.

I'm not sure why it being in the official contrib would matter. It's the official way to do GIS on PostgreSQL, the subject of multiple books, and the industry standard for any implementation of GIS for any database. It's also packaged by every major distribution and referenced in the docs.

* https://www.postgresql.org/download/products/6-postgresql-extensions/

**Comment by jbg** [comment]
Well, that's up to @sfackler, it's just a difference from the existing types that are implemented here (which as far as I can see are all part of PostgreSQL itself). FWIW, we're using `postgis` crate since quite a while, it works well, and the issues you identified are pretty minor API style things which I'm sure they will accept your PRs for.

**Comment by bjornharrtell** [comment]
I'm interested in this too. It's unclear to me how it could be externalized to another crate because to be truly useful.

In my case I found myself blocked on using `BinaryCopyInWriter` which is the most optimal way to bulk load data into PostgreSQL but I cannot use it for PostGIS geometry type, since that does not exist as a `postgres_types::Type`.

@sfackler what is your opinion here, are you interested in including first class PostGIS support in this crate?

**Comment by sfackler** [comment]
You can retrieve a `Type` instance by preparing a query that uses the type, and pulling it out of the statement: https://docs.rs/postgres/latest/postgres/struct.Statement.html#method.params.

Extension types do not have fixed Oids, and so cannot have hardcoded `Type` values.

**Comment by bjornharrtell** [comment]
Thanks @sfackler, that might be the key information I needed. 🙂 And might mean it does fit well in an external crate.

**Comment by bjornharrtell** [comment]
I can get the type which is fx. `Other { name: "geometry", oid: 4787952, kind: Simple, schema: "public" }` but I fail to understand how I can provide the ToSql conversion logic for such a Type. Probably only a lack of current insight on my part, but can't hurt to ask if it should be possible to provide customized ToSql for the Other type?

**Comment by sfackler** [comment]
Enable the `with-geo-types-0_7` Cargo feature and use https://docs.rs/geo-types/latest/geo_types/geometry/enum.Geometry.html.

**Comment by bjornharrtell** [comment]
I don't think that is what I need/want. What I want is fairly exotic perhaps, but it's specifically to be able to use the postgis geometry type as target for BinaryCopyInWriter. I have custom code that can produce the *native binary* representation of the postgis geometry type as a &[u8] and want to provide that as input to the write fn of BinaryCopyInWriter. That gives me this error:

Error { kind: ToSql(1), cause: Some(WrongType { postgres: Other(Other { name: "geometry", oid: 4787952, kind: Simple, schema: "public" }), rust: "&[u8]" }) }', src/main.rs:75:56

I intentionally do not want to go through another representation.

**Comment by sfackler** [comment]
If you already have the encoded bytes, then you can just make a ToSql implementation that directly writes them:

```rust
fn to_sql(&self, _: &Type, out: &mut BytesMut) -> Result<IsNull, Box<dyn Error + Sync + Send>> {
    out.extend_from_slice(&self.data);
    Ok(IsNull::No)
}
```

**Comment by bjornharrtell** [comment]
Might yet again be my lack of Rust experience that shows but I simply fail to understand how I can fit that snippet of code to apply to the moving parts here (the Other Type and &[u8]).

**Comment by sfackler** [comment]
```rust
struct MyGeometryWrapper {
    binary_encoded_geometry: Vec<u8>,
}

impl ToSql for MyGeometryWrapper {
    fn to_sql(&self, _: &Type, out: &mut BytesMut) -> Result<IsNull, Box<dyn Error + Sync + Send>> {
        out.extend_from_slice(&self.data);
        Ok(IsNull::No)
    }

    fn accepts(&self, ty: &Type) -> bool {
        ty.name() == "geometry"
    }

    postgres::to_sql_checked!();
}

let mut binary_copy_in_writer = BinaryCopyInWriter::new(copy_in_writer, &[geometry_type]);

let binary_encoded_geometry = encode_geometry_to_postgres_format();

binary_copy_in_writer.write(&[&MyGeometryWrapper { binary_encoded_geometry });
binary_copy_in
```

**Comment by bjornharrtell** [comment]
Much appreciated @sfackler, I got it working (after some tweaks) in my work in progress code to import osm data straight into postgis geometries.

A general implementation of the reusable parts should probably be contributable to the postgis crate, I hope to be able to do that at some point in the future.

**Comment by jjcfrancisco** [comment]
Hello, I am trying to implement this snippet but struggling to understand how do I pull the `geometry_type` part. I tried preparing a query but params comes out empty.

Help with this would be much appreciated.

**Comment by HendrikJanssen** [comment]
I had the same problem and ended up selecting the type oid from the database and constructing the type manually

First pull out the OID from the database:
```sql
SELECT oid FROM pg_type WHERE typname = 'geography'; -- Or !geometry'!
```

Then construct the type:

```rust
let geog_type = Type::from(String::from("geography"), <oid from select>, Kind::Simple, String::from("<schema>"));
```

This worked pretty good and let me use the BinaryCopyInWriter.

**Comment by EvanCarroll** [comment]
Well, update. I used the `postgis` crate without problem for years. Forgot about this. Forgot this was what I was doing. Tried to move to `geozero` and ran into **ALL THE SAME ERRORS** again.

**Comment by michaelkirk** [comment]
To restate my understanding:

The existing conversions in the postgres crate convert between geo-types and postgres's built in geometry types.

Most postgres users (like @EvanCarroll) who do GIS are using PostGIS's [geometry/geography](https://postgis.net/documentation/faq/geometry-or-geography/) types, which are not the same as postgres's built in geometry types.

It's likely that GIS folks (like @EvanCarroll) will never actually want to make use of the postgres <-> geo-types conversion, and by doing so, they are making a mistake.

Still, if Postgres has geometric types, it seems reasonable that we should be able to interact with them in rust, and that geo-types is a pretty reasonable rust representation for them.

So what's a reasonable resolution? Do we just need to have a better publicized "royal road" for GIS people to go down? 

I could understand if the `postgres` crate didn't want to include features for the third party PostGIS extension, but maybe we could at least help users like @EvanCarroll by at least linking to the solution they're actually looking for. 

I'm not sure what solution should be linked to, is it the [postgis](https://crates.io/crates/postgis) crate? I've never used it.

**Comment by EvanCarroll** [comment]
The idea solution is to implement actual GIS types in `geo-types`. Because without them we're not addressing the geospatial needs of the ecosystem. Currently, `geo-types` doesn't support _anything_ geospatial (srid) related. The `postgis` crate gives us a bouncing off point. I've written to author for maintenance on it. It seems like it's unmaintained. I wouldn't consider it in its current form for prime time, but as-is it's the only solution that supports types with an srid.

A lot of these problem imho stem from architectural decisions in geo-types and should be fixed there. I'm not sure how exactly that should occur, but one such idea would be to namespace
* `geo-types::*` into `geo-types::geometric`
* deprecate `geo-types::geometric::Point`
* create a new `geo-types::geospatial::Point` with srid support as `Option<u32>`.

After we have a `geo-types::geospatial::Point` the `postgis` crate or something that comes after it should NewType `geo-`types::geospatial::Point`into a `geographic` (srid default 4326) and `geometric` (srid default None). These should look up the oid in the `pg_catalog.pg_types` system, and follow the same general flow as other types in `postgresql-types` crate.

The geo-types crate already has the guts for `serde` support, so carrying that through to a ``geo-types::geospatial::Point`` a value add for all downstream applications (like a `geography` and `geometry`).
